nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/otp/OtpService.java

```diff
@@ -45,8 +45,12 @@
     // protected for testing purposes
     protected static final int MAX_CACHE_SOFT_LIMIT = 100;
 
-    private final Cache<CacheKey, String> downloadTokenCache;
-    private final Cache<CacheKey, String> uiExtensionCache;
+    private final Cache<CacheKey, String> downloadTokensToUsers;
+    private final Cache<CacheKey, String> uiExtensionTokensToUsers;
+
+    // keep a reverse cache to allow look-ups in both directions
+    private final Cache<String, CacheKey> usersToDownloadTokens;
```

Comment:
I think introducing parallel reverse-index caches adds more complexity than is necessary. This can lead to race conditions and concurrency problems. My understanding is that the only reason for the reverse-indexing is to allow retrieval of the user (key of the primary cache) by token (value of the primary cache). As the size of the cache is limited to 100 entries, I think it is reasonable to implement a `findByValue(String value)` method which checks for value presence and iterates to find the associated key. This can either be a static method which accepts the cache and token value as parameters, or the cache implementation can be extended with this instance method.

Reply:
I think you're right, adding the cache could expose the possibility for issues around concurrency. At the moment I'm busy doing some other things, if you'd like to instead try implementing this with an iterative check you can reassign this ticket. Thanks Andy!

Reply:
Will do.